[#911] Implement FDE status

sharvilshah opened the pull request:
##### TODO:

- [x] ~~Implement `encrypted` column for OS X in `block_devices`~~
- [x] Move OS X is_encrypted to a separate virtual table

(These are now tracked in #933)
* ~~Linux implementation~~
* ~~Add/update Tests~~
* ~~Documentation~~
 
Hey @theopolis,

This change adds `encrypted` column to the `block_devices` table.

Turns out `CoreStorage` is sub-class of  `IOMedia` (not a surprise), so this is a straightforward change, since there is already a `DASession`.

```bash
❯ echo "select name, uuid, encrypted from block_devices where encrypted == 1;"  \
| ./osqueryi

+------------+--------------------------------------+-----------+
| name       | uuid                                 | encrypted |
+------------+--------------------------------------+-----------+
| /dev/disk1 | ABCF88F5-629A-4DE9-9D30-91EFA68149E8 | 1         |
+------------+--------------------------------------+-----------+
```

```bash
❯ echo "select name, uuid, encrypted from block_devices where encrypted == 1;" \
 | ./osqueryi

+--------------+--------------------------------------+-----------+
| name         | uuid                                 | encrypted |
+--------------+--------------------------------------+-----------+
| /dev/disk0   |                                      | 0         |
| /dev/disk0s1 | 9E103246-DCFF-4ED7-ACA0-60F81F0A3E2A | 0         |
| /dev/disk0s2 | 107602D7-927A-47D7-B59E-9A72C7E47742 | 0         |
| /dev/disk0s3 | A3BBDEE1-437A-4C52-93D4-CFB9D1F6B4F6 | 0         |
| /dev/disk2   |                                      | 0         |
+--------------+--------------------------------------+-----------+
```

##### Questions:

* The UUIDs are logical volume ones, and not the LVF one (see #911 discussion), does it matter here?
* Since sqlite doesn't have a BOOLEAN type, what is the standard for the boolean valued columns? INTEGER 1 (true) and 0 (false) or TEXT true or false? I saw both being used in the table specs, I went with the INTEGER.
* Do we want to another another table for encryption status if we add encryption type and recovery key information? Or just add columns to `block_devices`? Or is `block_devices` already getting too cluttered?

osqueryer commented:
Can one of the osquery admins reply with "ok to test" to kick off a build...

theopolis commented:
Nice! I'd recommend a completely separate table for disk encryption using the BSDName/`name` as a foreign key from `block_devices`. You can add another volume_uuid, or lv_uuid, or just uuid to that table and include the algorithm/mode details. This way `block_devices` stays simple and portable. I'd hope the disk encryption table is portable, but that's not as important. 

Booleans, yeah use an INTEGER 1, 0 for now.

theopolis commented:
@sharvilshah what do you think of my suggestion above?

sharvilshah commented:
Yeah, I agree. I too prefer it as another virtual table instead of cluttering `block_devices`, and it's great that `FOREIGN` keys are supported.

For `encryption/cipher type`, I still haven't been able to find any references to it in `libcsfde.dylib` and I am thinking of hard coding "AES-XTS" which is the only thing filevault supports. Hard coding isn't ideal but I have spent too much time staring at disassembler output and *done is better than perfect*. 
This should be easier to accomplish for Linux, thoughts @theopolis?

theopolis commented:
I agree, AES-XTS hardcoded for now works. I'll ask around for cipher/key/alg details, open a developer support case, and submit an IT-help support ticket with Apple for advice. Usually we're asking for seemingly-irrelevant details but it's async so doesn't hurt.

sharvilshah commented:
Hey @theopolis,

Implemented FDE status for OS X as a separate table, can you please review? 
I am new to C++ and OS X APIs and might not be writing idiomatic c++ code, would appreciate a closer look. 

Here's how `fde_status` look:
```bash
+--------------+--------------------------------------+-----------+---------+
| name         | uuid                                 | encrypted | type    |
+--------------+--------------------------------------+-----------+---------+
| /dev/disk0   |                                      | 0         |         |
| /dev/disk0s1 | 9E103246-DCFF-4ED7-ACA0-60F81F0A3E2A | 0         |		    |
| /dev/disk0s2 | 107602D7-927A-47D7-B59E-9A72C7E47742 | 0         |		    |
| /dev/disk0s3 | A3BBDEE1-437A-4C52-93D4-CFB9D1F6B4F6 | 0         |		    |
| /dev/disk1   | ABCF88F5-629A-4DE9-9D30-91EFA68149E8 | 1         | AES-XTS |
| /dev/disk2   |                                      | 0         |         |
| /dev/disk2s1 |                                      | 0         |         |
| /dev/disk2s2 |                                      | 0         |         |
| /dev/disk3   |                                      | 0         |         |
| /dev/disk3s2 |                                      | 0         |         |
| /dev/disk3s1 |                                      | 0         |         |
+--------------+--------------------------------------+-----------+---------+
```

And here's the output of `profile.py`:

```
 D:0  C:0  M:0  F:0  U:0  fde_status (1/10): duration: 0.519093036652 cpu_time: 0.029035849 memory: 5545984 fds: 5 utilization: 2.75
 D:0  C:0  M:0  F:0  U:0  fde_status (2/10): duration: 0.521152019501 cpu_time: 0.024335588 memory: 5521408 fds: 5 utilization: 2.25
 D:0  C:0  M:0  F:0  U:0  fde_status (3/10): duration: 0.524597883224 cpu_time: 0.025143173 memory: 5550080 fds: 5 utilization: 2.25
 D:0  C:0  M:0  F:0  U:0  fde_status (4/10): duration: 0.519922971725 cpu_time: 0.024922294 memory: 5517312 fds: 5 utilization: 2.25
 D:0  C:0  M:0  F:0  U:0  fde_status (5/10): duration: 0.526874065399 cpu_time: 0.025863536 memory: 5513216 fds: 5 utilization: 2.35
 D:0  C:0  M:0  F:0  U:0  fde_status (6/10): duration: 0.519066810608 cpu_time: 0.025164828 memory: 5525504 fds: 5 utilization: 2.3
 D:0  C:0  M:0  F:0  U:0  fde_status (7/10): duration: 0.519506931305 cpu_time: 0.025122292 memory: 5500928 fds: 5 utilization: 2.3
 D:0  C:0  M:0  F:0  U:0  fde_status (8/10): duration: 0.521170139313 cpu_time: 0.025439418 memory: 5509120 fds: 5 utilization: 2.3
 D:0  C:0  M:0  F:0  U:0  fde_status (9/10): duration: 0.52365899086 cpu_time: 0.02467286 memory: 5513216 fds: 5 utilization: 2.2
 D:0  C:0  M:0  F:0  U:0  fde_status (10/10): duration: 0.520589113235 cpu_time: 0.02635973 memory: 5570560 fds: 5 utilization: 2.4
 D:0  C:0  M:0  F:0  U:0  fde_status   avg: duration: 0.521563196182 cpu_time: 0.0256059568 memory: 5526732.8 fds: 5.0 utilization: 2.335
```

Valgrind output:

```
❯ valgrind --dsymutil=yes --tool=memcheck --leak-check=yes --suppressions=./tools/tests/osquery.supp ./build/darwin/osquery/run 'SELECT * from fde_status;'
LEAK SUMMARY:
   definitely lost: 3,359 bytes in 48 blocks
   indirectly lost: 5,256 bytes in 20 blocks
     possibly lost: 15,160 bytes in 127 blocks
   still reachable: 105,060 bytes in 909 blocks
        suppressed: 362,334 bytes in 766 blocks
Reachable blocks (those to which a pointer was found) are not shown.
To see them, rerun with: --leak-check=full --show-leak-kinds=all
For counts of detected and suppressed errors, rerun with: -v
ERROR SUMMARY: 133 errors from 133 contexts (suppressed: 24 from 24)
```

I got the same output on my master (sha: 800dc77) with `SELECT * from block_devices;` query

If this looks good, I will get started on Linux implementation (need to read up on udev + dm-crypt), tests and docs.

##### Some more notes on FileVault

While tinkering some more with `libcsfde.dylib`, I came across `createNewPersonalRecoveryUser` procedure, so recovery key(s) are implemented by creating new users (not OS users), since FV2 supports multiple users. And that metadata + keys are stored encrypted probably in the `CoreStorage` header. This option doesn't seem worth the effort, and would be nice if Apple said there was some other API to figure out if recovery keys are set.

And realized Apple open source's some of darwin code, and found [headers/struct](https://opensource.apple.com/source/xnu/xnu-2782.1.97/libkern/crypto/corecrypto_aesxts.c) for xts: the function does take in a `cipher` param, but is ignored, much like in `IOKit` where sometimes we need to pass in `kNilOptions or 0`.

theopolis commented:
This looks great! 

I have some minor nits I'll leave inline. And it would be great if you could rebase to master then squash your local commits into a single body of work. 

Also, instead of fiddling with valgrind on OS X, check out Instruments' leaks. We have a simple wrapper using `./tools/profile.py --leaks --query "select * from table_name"`. The wiki page is a little buried, I should fix that soon: https://github.com/facebook/osquery/wiki/latency-and-performance-safety

theopolis commented on osquery/tables/system/darwin/fde_status.cpp:
```diff
@@ -0,0 +1,83 @@
+  }
+
+  auto property =
+      IORegistryEntryCreateCFProperty(service,
```
You can use our wrapper: [`getIOKitProperty`](https://github.com/facebook/osquery/blob/master/osquery/tables/system/darwin/iokit_utils.h#L19) if you're not bothered by a string comparison.

theopolis commented on osquery/tables/system/darwin/fde_status.cpp:
```diff
@@ -0,0 +1,83 @@
+  r["name"] = kDeviceNamePrefix + bsd_name;
+  r["uuid"] = uuid;
+  r["encrypted"] =
+      (property) ? CFBooleanGetValue((CFBooleanRef)property) ? "1" : "0" : "0";
```
`property` might be a `nullptr`, maybe move the check below (line 63) to about the row construction and only construct, push back, if the property has a value.

theopolis commented on osquery/tables/system/darwin/fde_status.cpp:
```diff
@@ -0,0 +1,83 @@
+  for (const auto& row : block_devices) {
+    auto name = row.at("name");
+    auto bsd_name = name.erase(0, kDeviceNamePrefix.size());
+    genFDEStatusForBSDName(bsd_name, row.at("uuid"), results);
```
The above two lines aren't speeding anything up, feel free to `row.at("name").substr(kDeviceNamePrefix.size())`

theopolis commented on third-party:
```diff
@@ -1 +1 @@
-Subproject commit 081abfe66d248457aa7f9b6498e3f65ee88cb5fe
```
You'll want to `git pull --rebase origin master`

theopolis commented on osquery/tables/specs/x/fde_status.table:
```diff
@@ -0,0 +1,11 @@
+table_name("fde_status")
```
Food for thought: I'm not 100% sold on this name, what do you think of: `disk_encryption` (doesn't fit with the mostly-plural existing names), `disk_encryption_status`, `disk_encryption_details`, `fde_details`? The reason I bring it up is there are a few instances where our not-so-standard acronyms may add confusion: `ad_config` (what does AD mean?), and `alf`, `application_firewall` would mean more.

It up to you. :) In this case "fde" is more common, so whatever name you think is most appropriate.

sharvilshah replied:
Yup `disk_encryption` sounds much better.

theopolis commented:
Example squash/rebase: https://github.com/theopolis/osquery/tree/filevault

sharvilshah commented:
Thanks @theopolis, updated and squashed my commits.

Ran the Instruments wrapper script:
```
Analyzing leaks in query: select * from disk_encryption
  definitely: 0 leaks for 0 total leaked bytes.
```

theopolis commented:
Mind squashing that final formatting fix too?

With respect to squashing: it's OK for a PR to have multiple commits, but they should represent distinct bodies of work. Thus if we want to revert a feature, bugfix, or change, we can do so with the GitHub UI or with a single commit. Squashing also provides a cleaner commit log, ours is a bit nasty littered with a ton of merge commits from the "merge pull request" workflow but we're working on that.

theopolis commented on osquery/tables/system/darwin/disk_encryption.cpp:
```diff
@@ -0,0 +1,79 @@
+/*
+ *  Copyright (c) 2015, Sharvil Shah
```
Also note that this will get code modded to include: Copyright (c) 2014, Facebook, Inc. as per the Facebook CLA.

sharvilshah replied:
I just modeled after one of the merged PR.

sharvilshah commented:
Makes sense. Squashed. Just needed to set up my git aliases.

theopolis commented:
Awesome! Want to merge this, or continue working on the next 3 parts within the same PR?

sharvilshah commented:
I already have a separate local branch for linux, will prefer separate PR. Will create a new issue too since #911 is too filevault specific.
